server/src/test/java/org/elasticsearch/snapshots/MetadataLoadingDuringSnapshotRestoreIT.java

```diff
@@ -0,0 +1,209 @@
+        RestoreSnapshotResponse restoreSnapshotResponse = client().admin().cluster().prepareRestoreSnapshot("repository", "snap")
+                                                                                    .setWaitForCompletion(true)
+                                                                                    .get();
+        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), equalTo(0));
+        assertThat(restoreSnapshotResponse.getRestoreInfo().status(), equalTo(RestStatus.OK));
+        assertGlobalMetadataLoads("snap", null);
+        assertIndexMetadataLoads("snap", "docs", 2);
+        assertIndexMetadataLoads("snap", "others", 2);
+
+        assertAcked(client().admin().indices().prepareDelete("docs"));
+
+        // Restoring a snapshot with selective indices loads only required index metadata
+        restoreSnapshotResponse = client().admin().cluster().prepareRestoreSnapshot("repository", "snap")
+                                                            .setIndices("docs")
+                                                            .setWaitForCompletion(true)
+                                                            .get();
+        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), equalTo(0));
+        assertThat(restoreSnapshotResponse.getRestoreInfo().status(), equalTo(RestStatus.OK));
+        assertGlobalMetadataLoads("snap", null);
+        assertIndexMetadataLoads("snap", "docs", 3);
+        assertIndexMetadataLoads("snap", "others", 2);
+
+        assertAcked(client().admin().indices().prepareDelete("docs", "others"));
+
+        // Restoring a snapshot including the global state loads it with the index metadata
+        restoreSnapshotResponse = client().admin().cluster().prepareRestoreSnapshot("repository", "snap")
+            .setIndices("docs", "oth*")
+            .setRestoreGlobalState(true)
+            .setWaitForCompletion(true)
+            .get();
+        assertThat(restoreSnapshotResponse.getRestoreInfo().failedShards(), equalTo(0));
+        assertThat(restoreSnapshotResponse.getRestoreInfo().status(), equalTo(RestStatus.OK));
+        assertGlobalMetadataLoads("snap", 1);
+        assertIndexMetadataLoads("snap", "docs", 4);
+        assertIndexMetadataLoads("snap", "others", 3);
+    }
+
+    private void assertGlobalMetadataLoads(final String snapshot, final Integer times) {
+        AtomicInteger count = getCountingMockRepository().globalMetadata.get(snapshot);
+        if (times == null) {
+            assertThat("Global metadata for " + snapshot + " must not have been loaded", count, nullValue());
+        } else {
+            assertThat("Global metadata for " + snapshot + " must have been loaded " + times + " times", count.get(), equalTo(times));
+        }
+    }
+
+    private void assertIndexMetadataLoads(final String snapshot, final String index, final Integer times) {
```

Comment:
use an int instead of an integer and use 0 to denote that indexmetadata was never loaded

Reply:
I wonder why I did this. Thanks